Sources/PubNub/EventEngine/Subscribe/SubscribeTransition.swift

```diff
@@ -0,0 +1,377 @@
+    from state: State,
+    error: PubNubError
+  ) -> TransitionResult<State, Invocation> {
+    return TransitionResult(
+      state: Subscribe.HandshakeFailedState(
+        input: state.input,
+        cursor: state.cursor,
+        error: error
+      ), invocations: [
+        .managed(.emitStatus(change: Subscribe.ConnectionStatusChange(
+          oldStatus: state.connectionStatus,
+          newStatus: .connectionError,
+          error: error
+        )))
+      ]
+    )
+  }
+}
+
+fileprivate extension SubscribeTransition {
+  func setReceivingState(
+    from state: State,
+    cursor: SubscribeCursor,
+    messages: [SubscribeMessagePayload] = []
+  ) -> TransitionResult<State, Invocation> {
+    let emitMessagesInvocation = EffectInvocation.managed(
+      Subscribe.Invocation.emitMessages(events: messages, forCursor: cursor)
+    )
+    let emitStatusInvocation = EffectInvocation.managed(
+      Subscribe.Invocation.emitStatus(change: Subscribe.ConnectionStatusChange(
+        oldStatus: state.connectionStatus,
+        newStatus: .connected,
+        error: nil
+      ))
+    )
+    let finalInvocations = [
+      !messages.isEmpty ? emitMessagesInvocation : nil,
+      state.connectionStatus != .connected ? emitStatusInvocation : nil
+    ].compactMap { $0 }
+    
+    return TransitionResult(
+      state: Subscribe.ReceivingState(input: state.input, cursor: cursor),
+      invocations: finalInvocations
+    )
+  }
+}
+
+fileprivate extension SubscribeTransition {
+  func setReceiveReconnectingState(
+    from state: State,
+    error: PubNubError
+  ) -> TransitionResult<State, Invocation> {
+    return TransitionResult(
+      state: Subscribe.ReceiveReconnectingState(
+        input: state.input,
+        cursor: state.cursor,
+        retryAttempt: ((state as? Subscribe.ReceiveReconnectingState)?.retryAttempt ?? -1) + 1,
+        reason: error
+      )
+    )
+  }
+}
+
+fileprivate extension SubscribeTransition {
+  func setReceiveFailedState(
+    from state: State,
+    error: PubNubError
+  ) -> TransitionResult<State, Invocation> {
+    guard let state = state as? Subscribe.ReceiveReconnectingState else {
+      return TransitionResult(state: state)
+    }
+    return TransitionResult(
+      state: Subscribe.ReceiveFailedState(
+        input: state.input,
+        cursor: state.cursor,
+        error: error
+      ), invocations: [
+        .managed(.emitStatus(change: Subscribe.ConnectionStatusChange(
+          oldStatus: state.connectionStatus,
+          newStatus: .disconnectedUnexpectedly,
+          error: error
+        )))
+      ]
+    )
+  }
+}
+
+fileprivate extension SubscribeTransition {
+  func setStoppedState(from state: State) -> TransitionResult<State, Invocation> {
+    let invocations: [EffectInvocation<Invocation>] = [
+      .managed(.emitStatus(change: Subscribe.ConnectionStatusChange(
+        oldStatus: state.connectionStatus,
+        newStatus: .disconnected,
```

Comment:
Handshake stopped transition shouldn't report disconnected because we are not subscribed.

Reply:
That's true, but there's still `connecting` status emitted when the user connects and the previous status wasn't active. What you mentioned is true and this should be removed after getting rid of `connecting` and `reconnecting` statuses from Swift SDK.